### ISSUE TITLE
(Feature request) Strike Through Font on Completed To-do Items

### ISSUE THREAD

**Issue by 0xbismarck** [zadam/trilium#4269]
### Describe feature

I would like to see a feature where I check an item complete on my to-do list, and Trilium crosses out the wording with strike through font.

**Comment by rauenzi** [comment]
Until it becomes a native feature you can use this:

```js
// Add to this array to add more classes to a completed to-do description
// You can select it in css with .todo-list__label__description
const classes = ["done"];

// Add more styles to this object to add more styles to completed to-do description
const styles = {
    "text-decoration": "line-through",
    "opacity": "0.7"
}

// Don't touch anything past here unless you know what you're doing!
const compiledStyles = Object.entries(styles).map(s => `${s[0]}: ${s[1]}`).join("; ");
function handleMutation(entries) {
    for (const entry of entries) {
        const added = entry.addedNodes[0];
        if (!added?.classList?.contains("todo-list__label")) continue;
        const checked = added.children[0].checked;
        const description = added.nextElementSibling;
        const classAction = checked ? "add" : "remove";
        description.classList[classAction](...classes);
        description.style = checked ? compiledStyles : "";
    }
}

const observer = new MutationObserver(handleMutation);
observer.observe(document, {childList: true, subtree: true});
```

Just add this to a JS Frontend note and give it `#run=frontendStartup` and it'll automatically start working when you start Trilium


https://github.com/zadam/trilium/assets/6865942/3718eb72-f9da-4f5d-9cbe-f18cc4bd50e6

**Comment by 0xbismarck** [comment]
Meant to post a follow-up that you're script works great. Thank you.

**Comment by 0xbismarck** [comment]
Playing around a little more with code, I notice when I navigate away from the note, and return the notes are no longer crossed out. Same issue when I close Trilium and reopen it.

**Comment by rauenzi** [comment]
It was written for a previous version of ckeditor, I will update the snippet if I have time

**Comment by JamesTann** [comment]
I wonder if this would also be possible using a JS event listener and some custom CSS, maybe in a note template?

**Comment by TheBig-O** [comment]
Were you able to sort out why it no longer works with the current version of ckeditor? Would love to have this functionality.
Thanks!

**Comment by herrkami** [comment]
@eliandoran, I personally distinguish between checked and struck-through todo items:
- [ ] Still to do
- [x] Done
- [x] ~~Abandoned/failed~~ -> didn't work out, but doesn't matter/consider it done
- [ ] ~~Abandoned/failed~~ -> didn't work out but still needs to be done/worked-around/replaced somehow

So this will (yet again 😭) break a very essential feature for me. I use todo notes and this particular feature literally every day, so I need to unset those CSS changes in my setup. Even if I started using them differently from now on, it would change my diary and previous notes semantically. 

I'd like to emphasize that I sincerely appreciate **all** the work you guys have been putting into trilium since triliumnext. Even this strike-through feature, really. But I'd like to use this opportunity to express my wish for the development to be a little more conservative/cautious when it overrides default functionality of modules or components. In this case, for example, check mark items lose a degree of freedom because two independent formatting functions (i.e. check and strikethrough) of CKEditor are implicitly merged into one. I totally see it's a fancy feature for some but it also means that checked items can't be formatted other than struck-through anymore. If you use checkmarks for a completed feature list, well bad luck then -- not only for me, I guess. 

Although this will likely be a fairly easy to fix for me, it's yet another fix I have to do. I'm certainly a very spleeny user, but recently I have to dedicate half a day per month on average only to keep my scripts and css halfway functional. Yet, many of my apps stopped working and I effectively lost quite a lot of functionality in Trilium over the course of the last year or so. 

Let me also emphasize that I love the pace at which trilium advances now. I certainly don't want to suggest slowing down the development or constraining it for legacy reasons. I think most of my pain points could be resolved very easily. Two examples:

If you'd ship the the strike-through feature (or similar features that are cool and neat for some but not actually part of Trilium's core functionality) as code notes, it would make it super easy to enable, disable, or customize them for users who'd like to do so. Zadam's Trilium had a lot of functionality implemented in this way (e. g. the entire To-Do system if I remember correctly) and the code notes are one of the most unique and awesome features of Trilium. Why not use those per default for little tweaks and improvements that are more on the gimmick side? I guess no new feature has been shipped as a code note since triliumnext, right? If so, why not? 

Another example is the integration of the move-line-up/down shortcut in text notes. Here, I think it's absolutely correct to implement this directly into trilium instead of deploying it as a code note. However, unlike (almost?) all other shortcuts, you cannot override it in Options/Shortcuts. Since both Ctrl + Up/Down, as well as Alt + Up/Down are dedicated to this function, those shortcuts became implicitly and silently unavailable for other purposes. It would have been very easy to avoid this by simply registering them in the shortcut options like all other the key combinations. 
Of course I know, I shouldn't talk about this other issue here. I should have created a new issue and you guys would have likely implemented it very very soon. But see, my time is limited and moving list items with shortcuts is another feature I'm using everyday since years. So I cannot wait for an issue to be solved, I need to solve it immediately because it violently disrupts my daily workflow. Once fixed, my motivation to raise an issue is also gone. For this, I'd like to apologize. 

To summarize my proposals/wishes: 
- Ship gimmicks as gimmicks (e.g. as code notes, as deselectable features, as configurable shortcuts, as new note types, etc.)
- Don't override default settings/functionality if it isn't necessary and think more carefully about whether changes might override or interfere with existing settings/functionality (also in the user space)
(New css development might break my theme but that's fine because that's the way it goes and I can simply change it with Trilium's theme system. Breaking a shortcut by hardcoding it outside the system that's supposed to make shortcuts configurable is, on the other hand, not fine imho.)

Sorry for the long comment and for going a bit off-topic.

**Comment by eliandoran** [comment]
> [@eliandoran](https://github.com/eliandoran), I personally distinguish between checked and struck-through todo items:
> 
> * [ ]  Still to do[x]  Done[x]  ~Abandoned/failed~ -> didn't work out, but doesn't matter/consider it done[ ]  ~Abandoned/failed~ -> didn't work out but still needs to be done/worked-around/replaced somehow
> 
> So this will (yet again 😭) break a very essential feature for me. I use todo notes and this particular feature literally every day, so I need to unset those CSS changes in my setup. 

This is a perfectly acceptable position. We've already provided a way to remove it, so no need to implement it from scratch: https://github.com/TriliumNext/Trilium/commit/d51e3de67474d11507ee433e4d373c89f5133a99#commitcomment-172298896

> 
> But I'd like to use this opportunity to express my wish for the development to be a little more conservative/cautious when it overrides default functionality of modules or components.

In principle the idea is good, but it's difficult to account for all the possible use cases the users might have. My position is that the benefit of the newly provided features outweigh the inconvenience of sometimes breaking stuff.

> In this case, for example, check mark items lose a degree of freedom because two independent formatting functions (i.e. check and strikethrough) of CKEditor are implicitly merged into one. I totally see it's a fancy feature for some but it also means that checked items can't be formatted other than struck-through anymore. If you use checkmarks for a completed feature list, well bad luck then -- not only for me, I guess.

The reasoning why we added strikethrough to the checkmarks is that it's a pretty common pattern in note taking application. Both Obsidian and Notion do the same. Taking inspiration from other applications is important because it makes the users feel more comfortable when switching to our app.

I know it can feel annoying to have features you might not personally need or want. Not all apps allow this kind of customization, so we didn't lose any functionality since it can be reverted. 

> Although this will likely be a fairly easy to fix for me, it's yet another fix I have to do. I'm certainly a very spleeny user, but recently I have to dedicate half a day per month on average only to keep my scripts and css halfway functional. Yet, many of my apps stopped working and I effectively lost quite a lot of functionality in Trilium over the course of the last year or so.

Can you tell me more about that? What are the custom implementations you have and how the changes in Trilium caused issues for you. If it's a big topic, we can have a call instead. 

> 
> Let me also emphasize that I love the pace at which trilium advances now. I certainly don't want to suggest slowing down the development or constraining it for legacy reasons. I think most of my pain points could be resolved very easily. Two examples:
> 
> If you'd ship the the strike-through feature (or similar features that are cool and neat for some but not actually part of Trilium's core functionality) as code notes, it would make it super easy to enable, disable, or customize them for users who'd like to do so. Zadam's Trilium had a lot of functionality implemented in this way (e. g. the entire To-Do system if I remember correctly) and the code notes are one of the most unique and awesome features of Trilium. Why not use those per default for little tweaks and improvements that are more on the gimmick side? I guess no new feature has been shipped as a code note since triliumnext, right? If so, why not?

There's a simple reason why we don't use code notes: there's no distribution mechanism. Scripts can only come from the demo, which is not updated in any way. Once we have a functional plugin system (which is a beast of its own), we can consider splitting them off. Without a central plugin registry, scripts are fragmented and thus it heavily impacts use (too much friction for most users to hunt scripts and CSS) 

> 
> Another example is the integration of the move-line-up/down shortcut in text notes. Here, I think it's absolutely correct to implement this directly into trilium instead of deploying it as a code note. However, unlike (almost?) all other shortcuts, you cannot override it in Options/Shortcuts. Since both Ctrl + Up/Down, as well as Alt + Up/Down are dedicated to this function, those shortcuts became implicitly and silently unavailable for other purposes. It would have been very easy to avoid this by simply registering them in the shortcut options like all other the key combinations.

This is not necessarily an issue with move blocks up/down, it's actually a problem with all the CKEditor keyboard shortcuts which are not actually changeable. 

> So I cannot wait for an issue to be solved, I need to solve it immediately because it violently disrupts my daily workflow. Once fixed, my motivation to raise an issue is also gone. For this, I'd like to apologize.

Yes, it's important that we are aware of the issues you are experiencing since otherwise there's no way to know if we broke something. It's also OK to write in the Matrix chat.

**Comment by herrkami** [comment]
Thanks again for your lightning response and the comprehensive explanations. Let me comment on a few points. 

> In principle the idea is good, but it's difficult to account for all the possible use cases the users might have. My position is that the benefit of the newly provided features outweigh the inconvenience of sometimes breaking stuff. 

I absolutely understand that this holds true for many things. But there are a few examples where I found, more flexible alternatives would have been very easy to implement. I'll list a few examples below. 

> Can you tell me more about that? What are the custom implementations you have and how the changes in Trilium caused issues for you. If it's a big topic, we can have a call instead.

I already created issues for my most urgent wishes. The two biggest are: 

1) CodeMirror mods are not possible anymore. However, I do not see how you could have implemented it any better. Upgrading CM was certainly the right thing to do and I don't think this should be considered as an example for what I meant above. 
However, due to this, I pretty much lost the ability to develop widgets within Trilium because too many editor features are missing now. Especially multi-cursor editing was a big thing for me. I also used it sometimes to batch edit content from text notes by copying over and back. I just list it here for completeness. 

2) Mobile launcher pane (let's call it MLP for now). I developed a weather widget in Trilium for which I had a shortcut in the MLP. I also had shortcuts to several checklists and scripts. For example, I had a button there to change from dark to bright theme without switching the theme of my entire phone. Or one that pasted timestamps into the text note (like Alt+T does on desktop). I understand why you decided to add another launcher pane menu in parallel to the existing one but I don't think it was necessary to remove the existing items as available options. That is more of a deliberate design decision, than a necessity, isn't it? And here, it was also pretty obvious that the change would break all existing shortcuts in mobile view without bringing significant functional improvement because the current button's functions were already available before the change. So in this case, I don't really recognize the outweighing benefits. 

3) This one is fixed for me but I'd like to mention it in more detail to illustrate better where my pain comes from: Move lines up/down shortcuts. The current implementation does not work perfectly for me because it doesn't allow moving lines up/down that are inside tables (CKE is a mess and I have many kinky spleens, yes.). Instead, it would move the entire table in this case. I often use tables to have several list categories right next to each other. Since you assigned both, Ctrl and Alt in a way that bypasses the Keyboard definitions in Options as well as the `api.bindGlobalShortcut()`, I assigned my own moving command to Ctrl + Alt + Up/Down which were still available. Since the shortcut assignment happens in Trilium, not in CKEditor, I disagree with it being an CKEditor problem. You could have hardcoded the command itself but bound the shortcut via the regular shortcut interface, exactly like it's done for pasting datetime or Trilium links, etc. Hardcoding _the shortcut_ outside the user's shortcut definitions is a deliberate decision that removes configurability and does not comply with how shortcuts are usually implemented in Trilium (i.e. _Options > Shortcuts > Text note operations_, or `api.bindGlobalShortcut()`). And also here, I find that potential conflicts with user shortcuts could have been anticipated without much headache and avoided without losing much time nor benefits. 

Concerning the plugins: Isn't the Options section implemented as Notes? They even have attributes and can be copied etc. Tbh, I've never taken a look into at how they're implemented but wouldn't it be possible to deploy code notes, in the same way? E.g. within a "Scripts" child note of Options? Users could copy them over to their user space to edit them there and they could enable/disable them with the regular attributes. 

Thank you for the offer to chat about it! I might join the matrix chat for topics that broader or more philosophical like this. I really appreciate the work and passion you put into this project -- please don't get me wrong, you're doing an amazing job here.

**Comment by GloogTs** [comment]
I'm using the Docker deployment and accessing it via the web interface. How can I disable the automatic strikethrough for completed To-do items?

**Comment by GloogTs** [comment]
> I'm using the Docker deployment and accessing it via the web interface. How can I disable the automatic strikethrough for completed To-do items?

I'm a non-technical user (no coding/CSS experience)

**Comment by eliandoran** [comment]
@GloogTs , please refer to this comment: https://github.com/TriliumNext/Trilium/commit/d51e3de67474d11507ee433e4d373c89f5133a99#commitcomment-172298896

It contains a link which explains how to create an App CSS.
